src/client/datascience/crossProcessLock.ts

```
@@ -0,0 +1,58 @@
+import { open, unlink } from 'fs-extra';
+import { tmpdir } from 'os';
+import * as path from 'path';
+import { sleep } from '../common/utils/async';
+
+export class CrossProcessLock {
+    private lockFilePath: string;
+    private acquired: boolean = false;
+
+    constructor(mutexName: string) {
+        this.lockFilePath = path.join(tmpdir(), `${mutexName}.tmp`);
+    }
+
+    public async lock(): Promise<boolean> {
+        const maxTries = 10;
+        let tries = 0;
+        while (!this.acquired && tries < maxTries) {
+            try {
+                await this.acquire();
+                if (this.acquired) {
+                    return true;
+                }
+                await sleep(100);
+            } catch (err) {
+                // Swallow the error and retry
+            }
+            tries += 1;
+        }
+        return false;
+    }
+
+    public async unlock() {
+        // Does nothing if the lock is not currently held
+        if (this.acquired) {
+            // Delete the lockfile
+            await unlink(this.lockFilePath);
+            this.acquired = false;
+        }
+    }
+
+    /*
+    One of the few atomicity guarantees that the node fs module appears to provide
+    is with fs.open(). With the 'wx' option flags, open() will error if the
+    file already exists, which tells us if it was already created in another process.
+    Hence we can use the existence of the file as a flag indicating whether we have
+    successfully acquired the right to create the keyfile.
+    */
+    private async acquire() {
+        try {
+            await open(this.lockFilePath, 'wx');
+            this.acquired = true;
+        } catch (err) {
+            if (err.code !== 'EEXIST') {
```

Not sure this is correct, should'nt we throw regardless (ignoring the error type).
E.g. if user cannot create file, then the lock was not acquired, and two processes can then have a conflict.

---

I say 1s is too low, cuz saving large notebooks can take more than 1s, & knowing how this will be used, we have at least 2 file writes that need to be performed by another process (write to digest store & write notebook)

---

I think its good to capture telemetry on the total time taken to acquire a lock as well. If > 1s, then we have some serious bottle neck.

---

Perhaps capture max duration in the last 15 minutes or so. this way we don't end up with 1000s of telemetry requests..

---

I.e not being able to create lock means, the lock was never acquired.

---

> Not sure this is correct, should'nt we throw regardless (ignoring the error type).

If the lockfile is opened with 'wx', it will be created if it doesn't already exist, and EEXIST will be thrown if it already exists. The lockfile gets deleted when the current process calls `unlock()`.

> I.e not being able to create lock means, the lock was never acquired.

If the current process tries to create the file and the file already exists, that means another process already holds the lock, so leave `this.acquired = false` because we failed to acquire the lock.

> knowing how this will be used, we have at least 2 file writes that need to be performed by another process (write to digest store & write notebook)

AFAIK we're only using this lock to guard key file creation, i.e. I don't think we are using this to guard writing to digest store and writing notebooks.

---

Agreed, came here to say that we don't need to worry about error conditions then saw your response.